demos/whisper-base/whisper.js

```diff
@@ -238,18 +341,54 @@ export class Whisper {
             attention_mask = new ort.Tensor("int32", new Int32Array(4).fill([1, 1, 1, 1]), [1, 4]);
         }
         // create decoder input for the first inference
-        const decoder_input = {
+        const decoder_inputs = {
             input_ids: new ort.Tensor("int32", new Int32Array(tokens), [1, 4]),
             attention_mask: attention_mask,
-            encoder_hidden_states: last_hidden_state,
+            encoder_hidden_states: encoder_outputs.last_hidden_state,
         };
         // console.log(`Non-KV cache decoder input preparation time: ${(performance.now() - start).toFixed(2)}ms`);
         // start = performance.now();
         // run the first inference which generates SA and CA KV cache
-        const decoder_output = await this.models["decoder"]["sess"].run(decoder_input);
+        // Create fetches object with pre-allocated outputs for IO binding
+        let logits, decoder_outputs;
+        if (this.ioBinding) {
+            // pre-allocated MLTensor outputs for IO binding
+            const first_logits = await this.createOutputMLTensor([1, 4, 51865], "float16", true);
+
+            // Create pre-allocated encoder key/value tensors for indices 0-5
+            const encoder_kv_tensors = {};
+            for (let i = 0; i < 6; i++) {
+                encoder_kv_tensors[`present_key_values.${i}.encoder.key`] = await this.createOutputMLTensor(
+                    [1, 8, 1500, 64],
+                    "float16",
```

Comment:
Is it more robust to use the data type based on the model data type, or a parameter or definition somewhere? I see `float32` used above for `createInputMLTensor(data, dims, dataType = "float32")`, and so I wonder when to use float16 vs when to use float32 🤔.

Reply:
Done and use `this.dataType` instead.